2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,383 @@
+  },
+  "Configuration": "[1.3.1,2.0)"
+}
+```
+
+>[!NOTE]
+>Credentials management will be using an upcoming RFC for generalized
+>credential management.  Not sure how to make that work in json other
+>than requiring json to have credential as a json object but that means
+>password/secret is in clear text.
+
+The `repository` property can be the name of the registered repository or the URL
+to a repository.
+If no `repository` is specified, it will use the `default` repository.
+
+The older `System.Version` four part version type will be supported to retain
+compatibility with existing published modules using that format.
+
+Due to the differences between semver and `System.Version`, we may have to keep
+the two distinct.
+
+Dependent declared modules not found on the system will prompt to be installed
+including module name, version, and size unless `-IncludeDependencies` is
+specified which will install without prompting.
+Declared dependencies are only searched within the same repository as the original
+module to be installed.
+
+### Saving resources
+
+This cmdlet uses `Install-PSResource -DestinationPath`.
+
+With the removal of PackageManagement, there is still a need to support saving
+arbitrary nupkgs (assemblies) used for scripts.
+
+`Save-PSResource -Type Library` will download nupkgs that have a `lib` folder.
+The dependent native library in `runtimes` matching the current system runtime
+will be copied to the root of the destination specified.
+A `-IncludeAllRuntimes` can be used to explicitly retain the `runtimes` directory
+hierarchy within the nupkg to the root of the destination.
+
+>[!NOTE]
+>Library support may not be available in 3.0.
+
+A `-Prerelease` switch allows saving prerelease versions.
+If the `-Version` includes a prerelease label like `2.0.0-beta4`, then this
+switch is not necessary and the prerelease version will be installed.
+
+A `-AsNupkg` switch will save the resource as a nupkg (if it was originally a
+nupkg) instead of expanding it into a folder.
+
+### Updating resources
+
+`Update-PSResource` will update all resources to most recent minor version by default.
+A `-AllowMajorVersionUpdate` switch will allow updating to newer major version.
+A `-PatchUpdateOnly` switch will allow updating only to patched versions.
+
+If the installed resource is a pre-release, it will automatically update to
+latest prerelease or stable version (if available).
+
+### Publishing resources
+
+`Publish-PSResource` will supporting publishing modules and scripts.
+It will follow the same model as `Publish-Module` in v2.
+
+A `-DestinationPath` can be used to publish the resulting nupkg locally.
+`Publish-PSResource` can accept a path to the nupkg to publish to a repository.
+
+A `-Nuspec` parameter can be used to specify a nuspec file rather than relying
+on this module to produce one.
+
+A `-SkipDependenciesCheck` switch can be used to bypass the default check that
+all dependencies are present.
+
```

Comment:
Except that this is like the stuff @markekraus was talking about above with package once, publish many -- if you sign it as part of publishing, then if you publish it to multiple locations, you have multiple different artifacts ...

Reply:
1. You could use caching to solve that.
1. If they don't solve that then each module/script author who signs and publishes needs to, whether they publish to multiple repos or not. Better to do the work once, upstream.
1. Publish could accept multiple targets (repo/cred pairs).

Reply:
1. You could use caching to solve that.
1. If they don't solve that then each module/script author who signs and publishes needs to, whether they publish to multiple repos or not. Better to do the work once, upstream.
1. Publish could accept multiple targets (repo/cred pairs).

Reply:
I don't think we'll have this in 3.0, but I'll add it to the considerations section as something we can address in the future.  Needs more thinking of the details of this.

Reply:
To support package once, publish many, it may make sense to break out package creation/build and publication into two separate cmdlets. A `New-PSResource` or `Build-PSResource` (`nuget.exe pack`) to create the `.nupkg` file and a `Publish-PSResource` (`nuget.exe push`) to publish to different repositories.